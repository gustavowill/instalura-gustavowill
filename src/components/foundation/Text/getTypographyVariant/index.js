import { css } from 'styled-components';
import breakpointsMedia from '../../../../theme/utils/breakpointsMedia';

export const TextStyleVariantsMap = {
  paragraph1: css`
    font-size: ${({ theme }) => theme.typographyVariants.paragraph1.fontSize};
    font-weight: ${({ theme }) => theme.typographyVariants.paragraph1.fontWeight};
    line-height: ${({ theme }) => theme.typographyVariants.paragraph1.lineHeight};
  `,
  paragraph1Bold: css`
    font-size: ${({ theme }) => theme.typographyVariants.paragraph1.fontSize};
    font-weight: 500;
    line-height: ${({ theme }) => theme.typographyVariants.paragraph1.lineHeight};
  `,
  paragraph2: css`
    font-size: ${({ theme }) => theme.typographyVariants.paragraph2.fontSize};
    font-weight: ${({ theme }) => theme.typographyVariants.paragraph2.fontWeight};
    line-height: ${({ theme }) => theme.typographyVariants.paragraph2.lineHeight};
  `,
  smallestException: css`
    font-size: ${({ theme }) => theme.typographyVariants.smallestException.fontSize};
    font-weight: ${({ theme }) => theme.typographyVariants.smallestException.fontWeight};
    line-height: ${({ theme }) => theme.typographyVariants.smallestException.lineHeight};
  `,
  title: css`
    ${({ theme }) => css`
      font-size: ${theme.typographyVariants.titleXS.fontSize};
      font-weight: ${theme.typographyVariants.titleXS.fontWeight};
      line-height: ${theme.typographyVariants.titleXS.lineHeight};
    `}
    ${breakpointsMedia({
    md: css`
        ${({ theme }) => css`
          font-size: ${theme.typographyVariants.title.fontSize};
          font-weight: ${theme.typographyVariants.title.fontWeight};
          line-height: ${theme.typographyVariants.title.lineHeight};
        `}
      `,
  })}
  `,
  subTitle: css`
    font-size: ${({ theme }) => theme.typographyVariants.subTitle.fontSize};
    font-weight: ${({ theme }) => theme.typographyVariants.subTitle.fontWeight};
    line-height: ${({ theme }) => theme.typographyVariants.subTitle.lineHeight};
  `,
};

export default function getTypographyVariant(VariantMap = TextStyleVariantsMap) {
  return ({ variant }) => {
    if (!variant) throw new Error('The component must have a variant');
    if (typeof variant === 'string') return VariantMap[variant];
    const breakpoints = Object.keys(variant);
    const typographyByBreakpoints = {};
    breakpoints.forEach((breakpoint) => {
      typographyByBreakpoints[breakpoint] = VariantMap[variant[breakpoint]];
    });

    return breakpointsMedia(typographyByBreakpoints);
  };
}
