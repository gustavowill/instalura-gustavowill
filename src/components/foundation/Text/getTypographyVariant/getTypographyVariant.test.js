import getTypographyVariant from '.';

const VariantsMapMock = {
  testVariant1: {
    fontSize: '12px',
    fontWeight: '400',
    lineHeight: '1',
  },
  testVariant2: {
    fontSize: '14px',
    fontWeight: '400',
    lineHeight: '1.25',
  },
  testVariant3: {
    fontSize: '16px',
    fontWeight: '500',
    lineHeight: '1.5',
  },
};

describe('getTypographyVariant()', () => {
  describe('when it is called in a component with a string variant', () => {
    test('returns the CSS for a single typography variant', () => {
      const internalFunc = getTypographyVariant(VariantsMapMock);
      const returnedCssObject = internalFunc({ variant: 'testVariant1' });
      expect(returnedCssObject).toMatchObject(VariantsMapMock.testVariant1);
    });
  });

  describe('when it is called in a component with breakpoint object variant', () => {
    test('returns the CSS for the listed breakpoints', () => {
      const internalFunc = getTypographyVariant(VariantsMapMock);
      const returnedCss = internalFunc({
        variant: {
          xs: 'testVariant1',
          md: 'testVariant2',
        },
      });
      expect(returnedCss).toMatchSnapshot();
    });
  });

  describe('when it is called in a component without a variant', () => {
    test('should throw a error', () => {
      const internalFunc = getTypographyVariant(VariantsMapMock);
      expect(() => internalFunc({})).toThrow(new Error('The component must have a variant'));
    });
  });
});
