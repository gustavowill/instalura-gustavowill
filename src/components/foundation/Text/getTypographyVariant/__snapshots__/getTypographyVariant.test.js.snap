// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTypographyVariant() when it is called in a component with breakpoint object variant returns the CSS for the listed breakpoints 1`] = `
Array [
  Array [
    "@media screen and (min-width:",
    "0",
    "px){",
    "font-size: 12px;",
    "font-weight: 400;",
    "line-height: 1;",
    "}",
  ],
  Array [
    "@media screen and (min-width:",
    "768",
    "px){",
    "font-size: 14px;",
    "font-weight: 400;",
    "line-height: 1.25;",
    "}",
  ],
]
`;
